lib/models/service.ts

```diff
@@ -0,0 +1,166 @@
+import * as DocumentDb from "documentdb";
+import * as DocumentDbUtils from "../utils/documentdb";
+import {
+  DocumentDbModelVersioned,
+  IVersionedModel,
+  ModelId
+} from "../utils/documentdb_model_versioned";
+
+import { Option } from "ts-option";
+import { Either } from "../utils/either";
+
+import { FiscalCode, isFiscalCode } from "../api/definitions/FiscalCode";
+
+import { nonEmptyStringToModelId } from "../utils/conversions";
+import { NonNegativeNumber } from "../utils/numbers";
+import { NonEmptyString } from "../utils/strings";
+
+/**
+ * Base interface for Service objects
+ */
+interface IBaseService {
+  // this equals user's subscriptionId
+  readonly serviceId: NonEmptyString;
+  // the name of the department within the service
+  readonly departmentName: NonEmptyString;
+  // the name of the service
+  readonly serviceName: NonEmptyString;
+  // the name of the organization
+  readonly organizationName: NonEmptyString;
+}
+
+/**
+ * Interface needed to interact with
+ * retrieved services from database
+ */
+export interface ISerializableService extends IBaseService {
+  // list of authorized fiscal codes
+  readonly authorizedRecipients?: ReadonlyArray<FiscalCode>;
```

Comment:
The problem with Sets that are not correctly serialized by `JSON.stringify` can be solved using [serializable Sets](https://www.npmjs.com/package/json-set-map) without having to define a separate `ISerializableService` type.

Reply:
this works if you use JSON.stringify() but we need the serialization part because all the DocumentDb methods work with plain json objects (ie. when you call a find() it returns a ISerializableService with an array and not a Set and that's the same with update() and create())

Reply:
the conversion from a DocumentDB document to our data type can be handled in the `toRetrieved` method defined in the model class

Reply:
what about the inverse conversion (to pass a value to update() and create()) ?

Reply:
you can pass the object `T`, the DocumentDB client relies on `JSON.stringify` to serialize the object, thus if `Set` gets correctly serialized you're good

Reply:
I'm trying to follow the path you've suggested but

Reply:
that's probably a bug, it should pass through `toRetrieved`

Reply:
what about this refactoring:

https://gist.github.com/gunzip/8718322cc7e7ab80161d6d1a16cb8b90

Reply:
LGTM